File: c++/back_tracking/q40.cpp
vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
    vector<vector<int>> ret;
    vector<int> unit;
    if(candidates.size()==0)
        return ret;
    sort(candidates.begin(),candidates.end());
    recur(ret,unit,candidates,target,0,0,candidates.size());
    return ret;
}

    void recur(vector<vector<int>>& ans, vector<int>& unit, vector<int>& arr, int target,int sum,int s, int e){
    if(sum>target)
        return;
    if(sum==target){
        ans.push_back(unit);
        return;
    }
    else{
        int i = s;
        while(i < e){
            unit.push_back(arr[i]);
            recur(ans,unit,arr,target,sum+arr[i],i+1,e);
            i += 1;
            unit.pop_back();
            while(i < e && arr[i]==arr[i-1])
                i+=1;
        }
    }

}

File: python/back_tracking/q40.py
def combinationSum2(self, candidates, target):
    """
    :type candidates: List[int]
    :type target: int
    :rtype: List[List[int]]
    """
    ret = []
    unit = []
    candidates.sort()
    self.dfs(candidates,ret,unit,target,0,)
    return ret

def dfs(self,arr,ret,unit,target,s):
    if target < 0:
        return
    if target == 0:
        ret.append(unit)
        return
    i = s
    while i < len(arr):
        self.dfs(arr,ret,unit+[arr[i]],target-arr[i],i+1)
        i += 1
        while i < len(arr) and arr[i] == arr[i-1]:
            i += 1
